README
Plack::App::Path::Router version 0.04
=====================================

See the individual module documentation for more information

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires these other modules and libraries:

    Plack
    Plack::Request
    Moose
    MooseX::NonMoose
    Path::Router

COPYRIGHT AND LICENCE

Stevan Little <stevan.little@iinteractive.com>

Copyright 2009-2011 Infinity Interactive, Inc.

http://www.iinteractive.com

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.



Changes
Revision history for Plack-App-Path-Router

{{$NEXT}}

0.08  2015-01-17

    - Change 'handle_exception' so that it rethrows exception by
      default, rather than returning it.

0.07 2015-01-13
    - Add 'handle_exception' method to
      Plack::App::Path::Router::Custom, to provide more granular
      control over response handling. (John Anderson)

    - Fix bug in component value extraction where a zero valued
      compotent would match but not extract properly. (John Anderson)

0.06  2014-09-01
    - return 404 via handle_response too
      (thanks to John "genehack" Anderson for this fix)

0.05  2012-04-20
    - Add Plack::App::Path::Router::Custom, which gives more control
      over exactly how things are passed into and out of the match target
    - Converted to Dist::Zilla

0.04  2011-01-12
    - Add Plack::App::Path::Router::PSGI, which allows you
      to dispatch to pure PSGI apps as targets
      - added the t/050_basic_psgi.t and t/051_basic_psgi_w_obj.t
        for this

0.03  2010-01-13
    - storing the Path::Router instance, match and other
      details in 'plack.router' instead of 'psgix.router'
      as per discussions with miyagawa on #plack

0.02  2010-01-07
    - adding support for custom request classes
      - added tests for this

0.01  2009-12-10
    - First release of this module.


lib/Plack/App/Path/Router.pm
package Plack::App::Path::Router;
use Moose 0.90;
use MooseX::NonMoose 0.07;
# ABSTRACT: A Plack component for dispatching with Path::Router

use Plack::Request 0.08;

extends 'Plack::App::Path::Router::Custom';

=head1 SYNOPSIS

  use Plack::App::Path::Router;
  use Path::Router;

  my $router = Path::Router->new;
  $router->add_route('/' =>
      target => sub {
          my ($request) = @_;
          # use the Plack::Request to
          # create a Plack::Response ...
          my $response = $request->new_response( 200 );
          $response->content_type('text/html');
          $response->body('<html><body>HELLO WORLD</body></html>');
      }
  );
  $router->add_route('/:action/?:id' =>
      validations => {
          id => 'Int'
      },
      target => sub {
          # matches are passed to the target sub ...
          my ($request, $action, $id) = @_;
          # return a PSGI response ...
          [
            200,
            [ 'Content-Type' => 'text/html' ],
            [ '<html><body>', $action, $id, '</body></html>' ]
          ]
      }
  );
  $router->add_route('/:action/edit/:id' =>
      validations => {
          id => 'Int'
      },
      target => sub {
          my ($r, $action, $id) = @_;
          # return a string (we will wrap
          # it in a PSGI response for you)
          "This is my action($action), and I am editing this id($id)";
      }
  );
  $router->add_route('/foo' =>
      # target objects are also supported
      # as long as the object responds to
      # the ->execute method
      target => MyApp::Action->new( type => 'FOO' )
  );

  # now create the Plack app
  my $app = Plack::App::Path::Router->new( router => $router );

=head1 DESCRIPTION

This is a L<Plack::Component> subclass which creates an endpoint to dispatch
using L<Path::Router>.

This module expects an instance of L<Path::Router> whose routes all have a
C<target> that is a CODE ref or an object which responds to the C<execute>
method. The CODE ref or C<execute> method will be called when a match is
found and passed a L<Plack::Request> instance followed by any path captures
that were found. It is expected that the target return one of the following;
an object which responds to the C<finalize> method (like L<Plack::Response>),
a properly formed PSGI response or a plain string (which we will wrap inside
a PSGI response with a status of 200 and a content type of "text/html").

This thing is dead simple, if my docs don't make sense, then just read the
source (all ~75 lines of it).

=cut

=attr router

This is a required attribute and must be an instance of L<Path::Router>.

=cut

=attr request_class

This is a class name used to create the request object. It defaults to
L<Plack::Request> but anything that will accept a PSGI-style C<$env> in
the constructor and respond correctly to C<path_info> will work.

=cut

has 'request_class' => (
    is      => 'ro',
    isa     => 'ClassName',
    default => sub { 'Plack::Request' },
);

has '+new_request' => (
    default => sub {
        my $self = shift;
        sub {
            $self->request_class->new(@_);
        };
    },
);

has '+target_to_app' => (
    default => sub {
        sub {
            my ($target) = @_;

            if (blessed $target && $target->can('execute')) {
                return sub { $target->execute(@_) };
            }
            else {
                return $target;
            }
        };
    },
);

has '+handle_response' => (
    default => sub {
        sub {
            my ($res) = @_;

            if ( blessed $res && $res->can('finalize') ) {
                return $res->finalize;
            }
            elsif ( not ref $res ) {
                return [ 200, [ 'Content-Type' => 'text/html' ], [ $res ] ];
            }
            else {
                return $res;
            }
        };
    },
);

__PACKAGE__->meta->make_immutable;
no Moose;

1;


lib/Plack/App/Path/Router/PSGI.pm
package Plack::App::Path::Router::PSGI;
use Moose 0.90;
use MooseX::NonMoose 0.07;
# ABSTRACT: A Plack component for dispatching with Path::Router to Pure PSGI targets

extends 'Plack::App::Path::Router::Custom';

=head1 SYNOPSIS

  use Plack::App::Path::Router::PSGI;
  use Path::Router;

  my $router = Path::Router->new;
  $router->add_route('/' =>
      target => sub {
          my $env  = shift;
          [
            200,
            [ 'Content-Type' => 'text/html' ],
            [ '<html><body>Home</body></html>' ]
          ]
      }
  );
  $router->add_route('/:action/?:id' =>
      validations => {
          id => 'Int'
      },
      target => sub {
          my $env = shift;
          # matches are passed through the $env
          my ($action, $id) = @{ $env->{'plack.router.match.args'} };
          [
            200,
            [ 'Content-Type' => 'text/html' ],
            [ '<html><body>', $action, $id, '</body></html>' ]
          ]
      }
  );
  $router->add_route('admin/:action/?:id' =>
      validations => {
          id => 'Int'
      },
      # targets are just PSGI apps, so you can
      # wrap with middleware as needed ...
      target => Plack::Middleware::Auth::Basic->wrap(
          sub {
              my $env = shift;
              # matches are passed through the $env
              my ($action, $id) = @{ $env->{'plack.router.match.args'} };
              [
                200,
                [ 'Content-Type' => 'text/html' ],
                [ '<html><body>', $action, $id, '</body></html>' ]
              ]
          },
          authenticator => sub {
              my ($username, $password) = @_;
              return $username eq 'admin' && $password eq 's3cr3t';
          }
      )
  );

  # now create the Plack app
  my $app = Plack::App::Path::Router::PSGI->new( router => $router );

=head1 DESCRIPTION

This is a L<Plack::Component> subclass which creates an endpoint to dispatch
using L<Path::Router>.

This module is similar to L<Plack::App::Path::Router> except that it expects
all the route targets to be pure PSGI apps, nothing more, nothing less. Which
means that they will accept a single C<$env> argument and return a valid
PSGI formatted response.

This will place, into the C<$env> the router instance into 'plack.router',
any valid match in 'plack.router.match' and the collected URL match args in
'plack.router.match.args'.

This thing is dead simple, if my docs don't make sense, then just read the
source (all ~45 lines of it).

=cut

=attr router

This is a required attribute and must be an instance of L<Path::Router>.

=cut

has '+target_to_app' => (
    default => sub {
        sub {
            my ($target) = @_;

            return blessed $target && $target->can('to_app')
                ? $target->to_app
                : $target;
        };
    },
);

__PACKAGE__->meta->make_immutable;
no Moose;

1;


lib/Plack/App/Path/Router/Custom.pm
package Plack::App::Path::Router::Custom;
use Moose 0.90;
use MooseX::NonMoose 0.07;
use Try::Tiny;
# ABSTRACT: A Plack component for dispatching with Path::Router

extends 'Plack::Component';

=head1 SYNOPSIS

  use Plack::App::Path::Router::Custom;
  use Path::Router;

  my $router = Path::Router->new;
  $router->add_route('/' =>
      target => sub {
          my ($request) = @_;
          # use the Plack::Request to
          # create a Plack::Response ...
          my $response = $request->new_response( 200 );
          $response->content_type('text/html');
          $response->body('<html><body>HELLO WORLD</body></html>');
      }
  );
  $router->add_route('/:action/?:id' =>
      validations => {
          id => 'Int'
      },
      target => sub {
          # matches are passed to the target sub ...
          my ($request, $action, $id) = @_;
          # return a PSGI response ...
          [
            200,
            [ 'Content-Type' => 'text/html' ],
            [ '<html><body>', $action, $id, '</body></html>' ]
          ]
      }
  );
  $router->add_route('/:action/edit/:id' =>
      validations => {
          id => 'Int'
      },
      target => sub {
          my ($r, $action, $id) = @_;
          # return a string (we will wrap
          # it in a PSGI response for you)
          "This is my action($action), and I am editing this id($id)";
      }
  );
  $router->add_route('/foo' =>
      # target objects are also supported
      # as long as the object responds to
      # the ->execute method
      target => MyApp::Action->new( type => 'FOO' )
  );

  # now create the Plack app
  my $app = Plack::App::Path::Router::Custom->new(
      router => $router,
      new_request => sub {
          my ($env) = @_;
          Plack::Request->new($env)
      },
      target_to_app => sub {
          my ($target) = @_;
          blessed($target) && $target->can('execute')
              ? sub { $target->execute(@_) }
              : $target
      },
      handle_response => sub {
          my ($res, $req) = @_;
          if (!ref($res)) {
              return [ 200, [], [$res] ];
          }
          elsif (blessed($res) && $res->can('finalize')) {
              return $res->finalize;
          }
          else {
              return $res;
          }
      },
  );

=head1 DESCRIPTION

This is a L<Plack::Component> subclass which creates an endpoint to dispatch
using L<Path::Router>.

It is useful when you need a bit more control than is provided by
L<Plack::App::Path::Router> or L<Plack::App::Path::Router::PSGI> (those two
modules are in fact written in terms of this one). It provides hooks to
manipulate how the PSGI env is turned into a request object, how the target is
turned into a coderef which accepts a request and returns a response, and how
that response is turned back into a valid PSGI response.

By default, the target must be a coderef which accepts a valid PSGI env and
returns a valid PSGI response.

=cut

=attr router

This is a required attribute and must be an instance of L<Path::Router>.

=cut

has 'router' => (
    is       => 'ro',
    isa      => 'Path::Router',
    required => 1,
);

=attr new_request

Coderef which takes a PSGI env and returns a request object of some sort.
Defaults to just returning the env.

=cut

has new_request => (
    traits  => ['Code'],
    isa     => 'CodeRef',
    default => sub { sub { $_[0] } },
    handles => {
        new_request => 'execute',
    },
);

=attr target_to_app

Coderef which takes the target provided by the matched path and returns a
coderef which takes a request (as provided by C<new_request>) and the match
arguments, and returns something that C<handle_response> can turn into a PSGI
response. Defaults to just returning the target.

=cut

has target_to_app => (
    traits  => ['Code'],
    isa     => 'CodeRef',
    default => sub { sub { $_[0] } },
    handles => {
        target_to_app => 'execute',
    },
);

=attr handle_response

Coderef which takes the response and request and returns a valid PSGI response.
Defaults to just returning the given response.

=cut

has handle_response => (
    traits  => ['Code'],
    isa     => 'CodeRef',
    default => sub { sub { $_[0] } },
    handles => {
        handle_response => 'execute',
    },
);

=attr handle_exception

Coderef which takes an exception thrown by the processing of the request and
returns a valid PSGI response. Defaults to just rethrowing the caught
exception.

=cut

has handle_exception => (
    traits  => ['Code'],
    isa     => 'CodeRef',
    default => sub { sub { die $_[0] } },
    handles => {
        handle_exception => 'execute' ,
    },
);


sub call {
    my ($self, $env) = @_;

    $env->{'plack.router'} = $self->router;

    my $req = $self->new_request( $env );

    my $match = $self->router->match( $env->{PATH_INFO} );

    if ( $match ) {
        $env->{'plack.router.match'} = $match;

        my $route   = $match->route;
        my $mapping = $match->mapping;

        my @args;
        foreach my $component ( @{ $route->components } ) {
            my $name = $route->get_component_name( $component );
            next unless $name;
            my $value = $mapping->{ $name };
            if (defined $value) {
                push @args => $value;
                $env->{ ('plack.router.match.args.' . $name) } = $value;
            }
        }

        $env->{ 'plack.router.match.args' } = \@args;

        my $target = $match->target;

        my $handled_response;

        try {
            my $app = $self->target_to_app( $target );
            my $res = $app->( $req, @args );

            $handled_response = $self->handle_response( $res, $req );
        }
        catch {
            $handled_response = $self->handle_exception( $_ );
        };

        return $handled_response;
    }

    return $self->handle_response(
      [ 404, [ 'Content-Type' => 'text/html' ], [ 'Not Found' ] ] ,
      $req ,
    );
}

__PACKAGE__->meta->make_immutable;
no Moose;

=head1 BUGS

All complex software has bugs lurking in it, and this module is no
exception. If you find a bug please either email me, or add the bug
to cpan-RT.

=cut

1;
